Report internal errors more clearly

sellout opened the pull request:
## Overview

Internal errors are rather opaque. In #4463, we see

``` ucm
scratch/main> run main

   renameTo: duplicate rename
```

this adds more context, so now we have

``` ucm
scratch/main> run main

  ❗️

  You’ve discovered an internal bug in the Unison runtime!

    renameTo: duplicate rename

  See if one of these issues at
  https://github.com/unisonweb/unison/issues reflects what
  you’re seeing. If not, please open a new one:
  * 3625
  * 4463
```

## Test coverage

There is a new transcript to replicate #4463, which does exercise this, but it will stop exercising it once that issue is fixed. But perhaps more reproductions will be added in the meantime …

I recommend reviewing commit-by-commit.

sellout committed:
Improve unison-runtime exception handling

- `CompileExn` and `internalBug` have been moved to a new module and had
  duplicate implementations removed,
- `internalBug` has a much richer message that should be more helpful to
  users,
- `internalBug` now accepts a list of issue numbers relevant to that
  message, and
- exception-related imports are now more explicit.

Also, there are a couple open questions tagged with __TODO__.

sellout committed:
Make `Pretty` `Width` more consistent

Previously there were three ways of handling optional `Width`

1. distinct `x` and `xUnbroken` functions,
2 `Maybe Width`, and
3 `Width 0` implies “unbroken”.

Since it’s difficult to use strictly positive numeric types[^1], this
now consistently treats `width <= 0`[^2] as “unbroken”.

[^1]: If it weren’t, I would have preferred something like `Maybe
      Positive`.

[^2]: It would be nice to define `Width` using something like `Word`
      instead of `Int`, which would allow `width == 0`, but that has
      cascading impacts across functions like `length` and `replicate`,
      which makes it more intrusive.

This also results in a couple inconsequential semantic changes to
“unbroken” rendering:

- “unbroken” is truly unbroken, not limited to `maxBound :: Int`
- right-aligning `Pretty` output (which isn’t currently used anyway)
  would do nothing in the “unbroken” case, rather than padding the
  output to `maxBound :: Int` columns.

sellout committed:
Add a transcript to reproduce #4463

sellout committed:
Improve messages for non-bug runtime exceptions

aryairani commented:
Hey this looks great. Two thoughts:

1. I think it'd be better if we could pass a super short description of what could be causing it, too:

```
scratch/main> run main

  ❗️

  Sorry — I've encountered a bug in the Unison runtime.

    renameTo: duplicate rename

  Please check if one of these known issues matches your situation:

  * repeated variable names in pattern match cause "duplicate rename" error
    https://github.com/unisonweb/unison/issues/4463

  If not, please open a new one:
  https://github.com/unisonweb/unison/issues/new/choose
```

Sorry I ended up making this a bad example by closing 3625 as a duplicate 😅 but yes it should take a list like you have already done.

2. We probably shouldn't set `runtime_tests_version` to your personal branch; can those changes be PRed into the runtime tests, and then we update here to 0.0.4?

sellout commented:
> 1. I think it'd be better if we could pass a super short description of what could be causing it, too:

Ah, yeah – I should have included this under “loose ends”. I was thinking we could fetch the title via the GitHub API, so the `internalBug` call sites still just have the issue number, but (as long as we don’t use `displayException`, which eschews IO) we could get the title dynamically.

> 2. We probably shouldn't set `runtime_tests_version` to your personal branch; can those changes be PRed into the runtime tests, and then we update here to 0.0.4?

Yeah, I was following the instructions in [interpreter-tests.md](https://github.com/unisonweb/unison/blob/trunk/unison-src/builtin-tests/interpreter-tests.tpl.md). I figured I shouldn’t open a contribution until this PR is stabilized (e.g., you suggestions will require an update to my runtime-tests branch).

sellout committed:
Update runtime error messages

- rephase
- include GitHub issue titles when possible

sellout committed:
Don’t redefine `die`

`Unison.Runtime.Exception.die` is overridden in `Machine.Types`, and
most call sites end up inheriting and using that instead. This just
folds the minor alteration of that version into the one in `Exception`.

sellout committed:
Merge remote-tracking branch 'upstream/trunk' into improve-internal-errors

aryairani commented:
> Ah, yeah – I should have included this under “loose ends”. I was thinking we could fetch the title via the GitHub API, so the `internalBug` call sites still just have the issue number, but (as long as we don’t use `displayException`, which eschews IO) we could get the title dynamically.

yeah that could work; we just have to remember to apply short github titles in that case

> Yeah, I was following the instructions in [interpreter-tests.md](https://github.com/unisonweb/unison/blob/trunk/unison-src/builtin-tests/interpreter-tests.tpl.md?rgh-link-date=2025-04-24T22%3A18%3A23Z). I figured I shouldn’t open a contribution until this PR is stabilized (e.g., you suggestions will require an update to my runtime-tests branch).

My bad. Ok so we'll stabilize the PR, then open the contribution for `/runtime-tests`, make a release, and update the PR, and then merge it?

sellout committed:
Merge remote-tracking branch 'upstream/trunk' into improve-internal-errors

sellout commented:
I think we had a race condition here – I pushed commits addressing your comment (it now pulls issue descriptions from GH), but you were writing another message at like the same time.

Anyway, I just merged in trunk and resolved conflicts, so hopefully this is in a good state to move forward with review.

aryairani commented:
Sorry @sellout, any ideas on the CI failures?

aryairani commented:
Ok, on the transcript test failures it looks like just a git merge issue; so they just need to be re-run and checked in again; for the interpreter test failures I'm not sure?

sellout commented:
> Sorry @sellout, any ideas on the CI failures?

Yeah. I think there was behavioral conflict when I merged in master – hashes in stack traces previously got resolved to names, but now they’re not. One of the transcript diffs is due to that too (I ran `transcripts` before updating this PR, but forgot to check `git diff` after).

aryairani commented:
I think we can use the "update transcripts" workflow to correct it automatically, but I can't seem to do it myself because the branch is in your repo.

sellout commented:
I think there’s actually a regression here, but I need to check a couple things. Wasn’t able to figure it out looking at the post merge version of the PR, so I think there’s a chance the regression is on trunk.

I’ll move this into draft until I sort that out.

aryairani commented:
erk, accumulating conflicts

sellout commented:
I’ll get this one wrapped today.

sellout committed:
Merge remote-tracking branch 'upstream/trunk' into improve-internal-errors

sellout committed:
Extract `prettyPanic` function

sellout commented:
Ok, this is finally passing. Before it can be merged,
1. https://share.unison-lang.org/@unison/runtime-tests/contributions/4 needs to be merged,
2. A new release of @unison/runtime needs to be cut, and
3. this PR needs to be updated to point at the new @unison/runtime-tests release instead of my branch.

dolio commented on .github/workflows/ci.yaml:
```diff
@@ -13,7 +13,7 @@ on:
 env:
   ## Some version numbers that are used during CI
-  runtime_tests_version: "@unison/runtime-tests/releases/0.0.3"
+  runtime_tests_version: "@unison/runtime-tests/@sellout/unison-5661"
```
I suspect we don't want to merge this?

sellout replied:
Yep, the new release was just cut today. Updated all occurrences.

dolio commented on unison-runtime/src/Unison/Runtime/Exception.hs:
```diff
@@ -1,34 +1,203 @@
+  = -- | pretty exception
+    PE CallStack [Word] (P.Pretty P.ColorText)
+  | -- | __TODO__: What is `BU`? Boxed/Unboxed?
+    BU [(Reference, Int)] Text Val
```
`BU` represents a call to something like `bug` in unison, with a unison stack, a message, and a unison value.

dolio commented on unison-runtime/src/Unison/Runtime/Stack.hs:
```diff
@@ -861,12 +861,15 @@ instance Show Stack where
+--
+--  __TODO__: Can this be represented with `These` instead, distinguishing the case where we don’t know which (`These`)
+--            from the known unboxed (`This`) and known boxed (`That`) cases? Or is it the case that we /do not/ know
+--            which?
```
This doesn't really have a `These` structure exactly. For instance, if the `UVal` is meaningful, the `BVal` tells you its type. So it's only in actual boxed cases that the `UVal` is meaningless. And I'm not certain that representing the cases like that will actually increase performance when these values are getting moved in and out of arrays.

sellout replied:
As an aside, I’ve found a lot of uses for a type like
``` haskell
data AndMaybe a b
  = Only a
  | Indeed a b
```
and this looks like another one. E.g.,
``` haskell
newtype Val = Val (!BVal `AndMaybe` !UVal)
```

sellout committed:
Merge remote-tracking branch 'upstream/trunk' into improve-internal-errors

sellout commented:
Ok, I merged in trunk again[^1], and this should now run CI against the most recent https://share.unison-lang.org/@unison/runtime-tests/contributions/4, which has had main merged in.

[^1]: I mostly merged just to get a fresh commit to trigger CI – resolving conflicts on this PR is pretty trivial, so I’m not necessarily trying to keep it current.

dolio reviewed (changes requested):
Mostly looks good (pretty mechanical). Added some explanations about todo comments, and maybe some oddities (although those might have been fixed since I started reviewing).

dolio commented on unison-src/builtin-tests/interpreter-tests.sh:
```diff
@@ -8,7 +8,7 @@ else
 fi
 
-runtime_tests_version="@unison/runtime-tests/releases/0.0.3"
+runtime_tests_version="@unison/runtime-tests/@sellout/unison-5661"
```
Should this be different as well?

aryairani replied:
I cut `releases/0.0.4` a little while ago.

sellout committed:
Point the runtime tests at the new release

sellout committed:
Replace “TODO” comments with real comments
